raft: introduce CommitMark

pav-kv opened the pull request:
This PR introduces the logically missing `CommitMark` coordinate system that carries the semantics of raft commits. A commit mark `(Term, Index)` guarantees that all the entries at indices <= `Index` in the log are committed as of the given `Term`, and will be present and committed in all future term logs.

These newly explained semantics are then used to formalize the meaning of the `raftLog.committed` field. Namely, the fact that `raftLog.committed` index is a `CommitMark` with `Term == raft.Term`.

The last commit moves the commit index capping logic inside `raftLog.commitTo`, and equips the latter with a safety check that ensures the commit index only moves forward if the reported `CommitMark` proves that out log matches the committed one. It also makes the commit index advancement more liberal, which today is not fully utilized and in the future can reduce tail latency of applies in leader change situations.

There is one case when this safe logic needs to be circumvented. When we receive a snapshot at the index that is already in the log, and the corresponding entry ID matches the snapshot, we [bump the commit index](https://github.com/cockroachdb/cockroach/blob/a590b2f8da8e086d421bf068d80b9ab621649420/pkg/raft/raft.go#L1889) unconditionally instead of applying the snapshot. This (pre-existing) case is what puts the commit index in the `CommitMark` coordinate system of the `raft.Term` leader, rather than our log's accepted term.

This quirk causes a "time travel" effect when the `accTerm` log can learn about an entry commit that actually happens at a higher term. The newly added comment for the `raftLog.committed` field explains in detail why this is safe: the phase-1 promise (`raft.Term`) prevents overwriting this entry by an uncommitted one. There is a second check that double-saves us: we [never try](https://github.com/cockroachdb/cockroach/blob/a590b2f8da8e086d421bf068d80b9ab621649420/pkg/raft/raft.go#L1715-L1718) to overwrite committed entries.

---

Part of #122100
Related to #127348, #127349
Related to #132030

pav-kv committed:
raft: introduce CommitMark

A commit mark (Term, Index) carries the guarantee that all the entries
at indices <= Index in the log are committed as of the given Term, and
will be present and committed in all future term logs.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132028)
<!-- Reviewable:end -->

pav-kv committed:
raft: make commitTo safe

This moves the commit index capping logic inside raftLog.commitTo, and
equips the latter with safety checks that ensure the commit index only
moves forward if the reported CommitMark belongs to the current or
preceding term leader.

There is one case when this logic needs to be circumvented. When we
receive a snapshot at the index that is already in the log, and the
corresponding entry ID matches the snapshot, we bump the commit index
unconditionally.

This (pre-existing) case puts the commit index in the CommitMark
coordinate system of the raft.Term leader, rather than our log's
accepted term. This enables the "time travel" effect when the accTerm
log can learn about an entry commit that actually happens at a higher
term. The newly added comment for the raftLog.committed field explains
in detail why this is safe.

Epic: none
Release note: none

pav-kv committed:
raft: explain the semantics of raftLog.committed

Epic: none
Release note: none

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -968,22 +968,13 @@ func (r *raft) appliedSnap(snap *pb.Snapshot) {
 	index := r.trk.Committed()
-	if index <= r.raftLog.committed {
-		// The commit index must not regress.
-		// TODO(pav-kv): consider making it an assertion.
```
NB: We can't make it an assertion because this inequality can happen, so I removed this TODO.

The reason why it can happen is that, initially, the leader doesn't know up-to-date `Match` indices for the followers. So, initially, `r.trk.Committed()` will return stale indices that may be behind `raftLog.committed`.

Could add a comment to this extent.

tbg replied:
I'd appreciate a comment.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 7 files at r1, 1 of 1 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132028#-:-O8cIoL8BcQ5XxN3WKhr:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @pav-kv, and @tbg)

___
*[`pkg/raft/log.go` line 382 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132028#-O8cH29p68oQyrQGYye0:-O8cH29p68oQyrQGYye1:b-15d60j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dce8cea7c1d5d2aceb32e5a5fd47f1577e3522d/pkg/raft/log.go#L382)):*
> ```Go
> }
> 
> // maybeCommit bumps the commit index if the given entry ID matches the log.
> ```

Mind mentioning what the return value of this function means?
___
*[`pkg/raft/log.go` line 382 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/132028#-O8cHbhO2qRKy3umYCQq:-O8cHbhO2qRKy3umYCQr:b-sd7n4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dce8cea7c1d5d2aceb32e5a5fd47f1577e3522d/pkg/raft/log.go#L382)):*
> ```Go
> }
> 
> // maybeCommit bumps the commit index if the given entry ID matches the log.
> ```

What can we do to better differentiate this function and `commitTo`? They play very different roles, and yet their names and comments make them seem quite similar.
___
*[`pkg/raft/types.go` line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132028#-O8cD_vS8i-uCKqcDFCs:-O8cD_vS8i-uCKqcDFCt:b6enz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/81b9f36622fbf0d64d5d9ab5cb3d9cbdcbc1bb7e/pkg/raft/types.go#L91)):*
> ```Go
> //     different proposals, up to a Term under which one of them gets committed.
> //   - If/when the entry is committed, it enters the CommitMark coordinate
> //     system, which is scoped to learners. Entries are transferred to the state
> ```

This raft library uses the term "learner" differently. Consider reminding readers here that we still mean the general concept in Paxos, and not the role with the same name in this library.


<!-- Sent from Reviewable.io -->

tbg commented on pkg/raft/log.go:
```diff
@@ -54,8 +54,28 @@ type raftLog struct {
+	//
+	// The invariant we rely on is: accTerm <= raft.Term. Raft has given a promise
+	// to never accept log writes from terms < raft.Term. Since the committed
+	// index is in raft.Term commit mark coordinate system, we have the guarantee
```
Is this "just" saying that a leader will never tell a follower about a committed index it doesn't already have in the log?

If so: I'm not sure whether "since the committed index is in raft.Term commit mark coordinate system" is a good way to express this. Something being in a coordinate system doesn't guarantee anything, right? It's that we only talk about something in this coordinate system once some conditions are fulfilled. I think the commit index coordinate system is useful, but if we lean on it too much to communicate our invariants, we might be obfuscating rather than explaining.

If not: whatever this means, I couldn't figure it out.

tbg commented on pkg/raft/log.go:
```diff
@@ -379,22 +379,44 @@ func (l *raftLog) lastIndex() uint64 {
+		// NB: in this case, our log is lagging the leader's. If the leader is
+		// stable, we will eventually accept a MsgApp or snapshot which bumps
+		// accTerm and enables advancing the commit index again. By this, we have
+		// the guarantee that our commit index converges to the leader's.
```
do we actually expect to hit this path? I'm not against adding a check here, but we should document whether this is basically an assertion failure, and consider logging an error.

tbg reviewed (commented):
I'm glad you're working on clarifying the semantics here and moving logic into one place - reviewing this was a humbling demonstration of how poorly I understand these subtle details.
Even with the additional comments, (admittedly in a travel-induced stupor but still) I don't think I've grokked everything. I've left comments on things that I struggled with, and think I'll need to take another pass. Either way, I can already tell this is much needed cleanup.

pav-kv commented on pkg/raft/log_unstable.go:
```diff
@@ -227,23 +227,26 @@ func (u *unstable) stableSnapTo(i uint64) {
+	// If mark.Term <= u.term, our log is consistent with the snapshot. Set the
+	// new accepted term to the max of the two so that u.term does not regress. At
+	// the time of writing, mark.Term is always >= u.term, because mark.Term ==
+	// raft.Term and u.term <= raft.Term. It could be relaxed in the future.
```
Good example.

> the snapshot will have index=lastIndex(recipient), term 3, and the unstable on the recipient will have term 4, no?

Yes, the snapshot will be at `entryID` = `{index:lastIndex, term:3}`, but the commit mark in this terminology is `{term:5, index:lastIndex}`.

So, at the recipient, who is at `LogMark.Term == 4` and, say, `Term=4`, we will overwrite the log with this newer-term committed state, and update `Term` to 5 and `LogMark.Term` to 5 (the `max` in the next line). Our log mark moves forward, and we're safe to regress to the snapshot's index as long as its last `entryID` does not match our log (which is checked up the stack in the caller). If it matches then there is a chance we have a suffix that must be retained.

---

Another example (impossible situation):

```
log (terms only)

        3 3 3 3   <- sender (all committed at term 3)
       /
3 3 3
       \
        4 4 4 4 4 4 4    <- recipient
```

In this case, the commit `mark` is `{term:3, index}`. Since these were committed at term 3, all term 4 logs must contain these entries. So we can't have this situation.

pav-kv commented on pkg/raft/log_unstable.go:
```diff
@@ -227,23 +227,26 @@ func (u *unstable) stableSnapTo(i uint64) {
 	// it, but at the time of writing the whole stack (including storage) is
 	// written such that a snapshot always clears the log.
-	if s.term <= u.term && s.lastIndex() < u.lastIndex() {
+	if mark.Term <= u.term && mark.Index < u.lastIndex() {
```
> 1. the snapshot contains a (Index, Term) pair. This pair can be understood as an entryID - the snapshot is the materialized state of the entry. This can also be understood as a commit mark - the entry was committed (or it couldn't be materialized in a snapshot)

Yes, the `(Index,Term)` is an `entryID`, but not a `CommitMark` in this terminology. Since it's number 1 in your list, see if it changes the chain of thought.

See the [snapshot](https://github.com/cockroachdb/cockroach/blob/645eb8c99796b3b88f5631aa0fc92a011010ce64/pkg/raft/types.go#L165-L190) wrapper which encompasses both the pb snapshot (`entryID`) and the term of its sender. For notation simplicitly, let's say that a snapshot (or any committed log prefix) is `entryID`@`term`.

Example committed log: `[2 2 2 2 2 3 3 3] @ term 3`. Say entries `[2 2 2 2 2]` were not committed under term 2, and only became committed under term 3 (when it first happened that raft's commit criterion "fired"). If so, any `CommitMark` that covers these `[2 ... 2]` entries will have `CommitMark.Term >= 3`; and any leader log at term >= `CommitMark.Term` will have all these entries (or any acceptor log at `accTerm` >= `CommitMark.Term`, which is the same thing). The existence of a `CommitMark` tells a partial story how this entry got committed, and communicates provenance.

> 2. for better or worse, we erase the raft log when applying a snapshot. So we must never apply one that would have us remove any log entries that the local raft instance may have acked as part of a quorum.

Yep, the challenge in raft is that the acceptor's log and the learner's state (snapshot + committed log) are intertwined, particularly we reason about "accepted state" of this combined node (in voting etc) as a `f(acceptor, learner)`. So we should somehow combine their coordinate systems to work together correctly.

Right now, I think, this combined "system" is (`Term`, `lastEntryID`, `commitIndex`). And this PR moves things a bit in the direction of some decoupling. Clearly needs more thought and scrutiny.

> 3. if the snapshot commit mark's term is (strictly) leading the unstable term (which is the entryID.term of the newest (stable or unstable) entry, the snapshot commits an entry we haven't ever seen, so it is safe to discard the log.

The last `entryID` in the log can have an older term than `unstable.term` (which is `LogMark.Term`). Say, if leader @ term 5 replicated a log `[2 2 3 3]` to us with the latest append, the last entry term is 3, but the `LogMark.Term` is 5.

It is safe to discard the log if one of of the following is true, for the `entryID`@`term` snapshot:
- We don't have `entryID` in our log. This means our log is stale compared to the committed state, and we can replace it.
- `entryID.index` >= our `lastIndex`. Same reason.

pav-kv reviewed (commented):
@tbg Replied to a couple of your comments to maybe trigger some more thoughts. I'm not going to merge this PR for now, before we have some shared understanding. Probably will create a doc some time later, with visuals/notation etc, where we could discuss more / converge; I'd start with trying to answer the questions that you raised here. Maybe there is a less obfuscated way to explain things, or a better "model".
